File: go/magic_number.go
package main

import (
	"errors"
	"fmt"
	"math"
	"strconv"
)

var errGetMN = errors.New("Error getting magic number")

type E struct {
	seed            int
	currentNumber   int
	offsetParameter int
	multiplier      int
}

func (E *E) initE(r, t, e int) {
	E.seed = r
	E.currentNumber = r % t
	E.offsetParameter = t
	E.multiplier = e
	if E.currentNumber <= 0 {
		E.currentNumber += t
	}
}

func (E *E) getNext() int {
	E.currentNumber = E.multiplier * E.currentNumber % E.offsetParameter
	return E.currentNumber
}

func inE(r string) int {
	t := 126 ^ int([]rune(r)[0])
	for e := 1; e < len(r); e++ {
		t += (int([]rune(r)[e])*e ^ int([]rune(r)[e-1])) >> (e % 2)
	}

	return t
}

type DatadomeMagicNumber struct {
	language  string
	languages []string
	r         string
	t         int
	userAgent string
}

func (ddmn *DatadomeMagicNumber) first(r, t int) int {
	e := 26157
	n := 0
	s := "VEc5dmEybHVaeUJtYjNJZ1lTQnFiMkkvSUVOdmJuUmhZM1FnZFhNZ1lYUWdZWEJ3YkhsQVpHRjBZV1J2YldVdVkyOGdkMmwwYUNCMGFHVWdabTlzYkc5M2FXNW5JR052WkdVNklERTJOMlJ6YUdSb01ITnVhSE0"
	for a := 0; a < len(s); a++ {
		ibin, _ := strconv.Atoi(strconv.FormatInt(int64([]rune(s)[a]), 2))
		n += ibin | (e ^ t)
	}

	return n
}

func (ddmn *DatadomeMagicNumber) second(r, t int) int {
	e := strconv.FormatInt(int64(len(ddmn.userAgent)<<int(math.Max(float64(r), 3))), 2)
	n := -42
	for a := 0; a < len(e); a++ {
		n += int([]rune(e)[a]) ^ t<<(a%3)
	}

	return n
}

func (ddmn *DatadomeMagicNumber) third(r, t int) int {
	fmt.Println(r, t)
	e := 0
	n := ddmn.language[:2] + strconv.Itoa(t)
	for a := 0; a < len(n); a++ {
		e += int([]rune(n)[a]) << int(math.Min(float64((a+t)%(1+r)), 2))
		e = (e << 3) - e + int([]rune(n)[a])
		e = (e & e) >> a
	}
	fmt.Println(e)

	return e
}

func (ddmn *DatadomeMagicNumber) Generate() int {
	n := []func(int, int) int{ddmn.first, ddmn.second, ddmn.third}
	a := &E{}
	inE := inE(ddmn.r)
	a.initE(inE, 1723, 7532)
	o := a.seed
	for u := 0; u < ddmn.t; u++ {
		o ^= n[a.getNext()%len(n)](u, a.seed)
	}

	return o
}

File: py/magic_number.py
import math


class e:
    def __init__(self, r, t, e):
        self.seed = r
        self.currentNumber = r % t
        self.offsetParameter = t
        self.multiplier = e
        if self.currentNumber <= 0:
            self.currentNumber += t

    def getNext(self):
        self.currentNumber = self.multiplier * self.currentNumber % self.offsetParameter
        return self.currentNumber


def inE(r):
    t = 126 ^ ord(r[0])
    e = 1
    while e < len(r):
        x = ord(r[e]) * e
        y = ord(r[e - 1])
        t += (x ^ y) >> e % 2
        e += 1
    return t


class DatadomeMagicNumber:
    def __init__(self, r, t, ua):
        self.language = "en-GB"
        self.languages = ["en-GB", "es", "en", "zh", "zh-TW", "zh-CN", "th", "nl", "ru"]
        self.r = r
        self.t = t
        self.userAgent = ua

    def first(self, r, t):
        e = 26157
        n = 0
        s = "VEc5dmEybHVaeUJtYjNJZ1lTQnFiMkkvSUVOdmJuUmhZM1FnZFhNZ1lYUWdZWEJ3YkhsQVpHRjBZV1J2YldVdVkyOGdkMmwwYUNCMGFHVWdabTlzYkc5M2FXNW5JR052WkdVNklERTJOMlJ6YUdSb01ITnVhSE0"
        a = 0
        while a < len(s):
            n += int(str(bin(ord(s[a])))[2:]) | e ^ t
            a += 1  #% math.ceil(4.1425172 / len("self.userAgent"))

        return n

    def second(self, r, t):
        e = bin(len(self.userAgent) << max(r, 3))[2:]
        n = -42
        a = 0
        while a < len(e):
            n += ord(e[a]) ^ t << a % 3
            a += 1

        return n

    def third(self, r, t):
        e = 0
        n = self.language[:2] + str(t)
        a = 0
        while a < len(n):
            e += ord(n[a]) << min((a + t) % (1 + r), 2)
            e = (e << 3) - e + ord(n[a])
            e = (e & e) >> a
            a += 1

        return e

    def Generate(self):
        n = [
            self.first,
            self.second,
            self.third,
        ]
        a = e(inE(self.r), 1723, 7532)
        o = a.seed
        u = 0
        while u < self.t:
            o ^= n[a.getNext() % len(n)](u, a.seed)
            u += 1

        return o
